### ISSUE TITLE
System.ArgumentNullException: "SafeHandle cannot be null" on every http request

### ISSUE THREAD

**Issue by KKhurin** [dotnet/corefx#9578]
Looks like we missed this in WCF HelloWorld perf/stress :( 

There is a System.ArgumentNullException per WCF request which gets handled and the request proceeds successfully, but the exception isn't the most performant way to control the execution flow ... 

I didn't dig into the details on why it happens (whether it needs a multi-threaded access to http client or not), I only have the call stack for now:
~~~
System.StubHelpers.StubHelpers.SafeHandleAddRef(System.Runtime.InteropServices.SafeHandle, Boolean ByRef)
Interop+WinHttp.WinHttpSetOption(SafeWinHttpHandle, UInt32, UInt32 ByRef, UInt32)
DomainBoundILStubClass.IL_STUB_PInvoke(SafeWinHttpHandle, UInt32, UInt32 ByRef, UInt32)
System.Net.Http.WinHttpHandler+d__101.MoveNext()
System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
System.Runtime.CompilerServices.AsyncMethodBuilderCore+MoveNextRunner.RunWithDefaultContext()
System.Threading.ThreadPoolWorkQueue.Dispatch()
~~~

**Comment by KKhurin** [comment]
@stephentoub @SajayAntony 
Confirmed, this is per every simple httpclient request:
~~~
new HttpClient().GetAsync("http://localhost/").Wait();
~~~

**Comment by stephentoub** [comment]
@KKhurin, this is Windows-only, right? (From the stack, I assume so.)

**Comment by stephentoub** [comment]
cc: @davidsh, @cipop

**Comment by stephentoub** [comment]
@davidsh, I took a quick look at this.  It's happening because CreateResponseMessage is called here:
https://github.com/dotnet/corefx/blob/cf0f5bf67212aabdfae216fac09081319c00fa18/src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs#L865
just before SetWinHttpOption is called with state.RequestHandle to set the receive timeout:
https://github.com/dotnet/corefx/blob/cf0f5bf67212aabdfae216fac09081319c00fa18/src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs#L871
but CreateResponseMessage nulls out state.RequestHandle.
https://github.com/dotnet/corefx/blob/cffe60fcec1f85f755f41562b0c207700fdbc043/src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpResponseParser.cs#L56
The exception is thrown and immediately caught, but then since there is a response message, the exception is ignored:
https://github.com/dotnet/corefx/blob/cf0f5bf67212aabdfae216fac09081319c00fa18/src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs#L889-L896

**Comment by SajayAntony** [comment]
@stephentoub this shows up as a significant cost in the middle tier scenario (like WCF) and I expect the pattern will apply to other stacks like the storage SDK. So would be worth fixing for 1.1 at least.

**Comment by stephentoub** [comment]
> this shows up as a significant cost 

I'm curious, do you have numbers?

> So would be worth fixing for 1.1 at least.

I agree, but David and his team own this.

**Comment by davidsh** [comment]
I will fix this today.  I broke this when I fixed #8996.  Thanks for pointing out the problem. Since the exception was being caught internally, this didn't show up in normal CI testing.

**Comment by erick-thompson** [comment]
Thanks for being responsive on this one. I was tracking down another exception in VS by breaking when thrown, and this was the source of a lot of noise.
